In [16]:
import pandas as pd
import numpy as np

In [17]:
df = pd.read_csv('AQI.csv')
df

City        Date   PM2.5    PM10     NO    NO2    NOx    NH3  \
0          Ahmedabad  2015-01-01   34.52  154.75   0.92  18.22  17.15   8.98   
1          Ahmedabad  2015-01-02   25.83  226.24   0.97  15.69  16.46   9.09   
2          Ahmedabad  2015-01-03   90.95  206.09  17.40  19.30  29.70   8.99   
3          Ahmedabad  2015-01-04   25.83  226.24   1.70  18.48  17.97   9.09   
4          Ahmedabad  2015-01-05  156.12  236.77  22.10  21.42  37.76   8.91   
...              ...         ...     ...     ...    ...    ...    ...    ...   
29526  Visakhapatnam  2020-06-27   15.02   50.94   7.68  25.06  19.54  12.47   
29527  Visakhapatnam  2020-06-28   24.38   74.09   3.42  26.06  16.53  11.99   
29528  Visakhapatnam  2020-06-29   22.91   65.73   3.45  29.53  18.33  10.71   
29529  Visakhapatnam  2020-06-30   16.64   49.97   4.05  29.26  18.80  10.03   
29530  Visakhapatnam  2020-07-01   15.00   66.00   0.40  26.85  14.05   5.20   

          CO    SO2      O3  Benzene  Toluene  Xylene     AQI    AQI_Bucket  
0       0.92  27.64  133.36     0.00     0.02    0.00  133.02      Moderate  
1       0.97  24.55   34.06     3.68     5.50    3.77  152.90      Moderate  
2      17.40  29.07   30.70     6.80    16.40    2.25  294.05          Poor  
3       1.70  18.59   36.08     4.43    10.14    1.00  154.82      Moderate  
4      22.10  39.33   39.31     7.01    18.89    2.78  397.48     Very Poor  
...      ...    ...     ...      ...      ...     ...     ...           ...  
29526   0.47   8.55   23.30     2.24    12.07    0.73   94.67  Satisfactory  
29527   0.52  12.72   30.14     0.74     2.21    0.38  108.25      Moderate  
29528   0.48   8.42   30.96     0.01     0.01    0.00  101.48      Moderate  
29529   0.52   9.84   28.30     0.00     0.00    0.00   93.67  Satisfactory  
29530   0.59   2.10   17.05     0.00     0.00    0.00   88.30  Satisfactory  

[29531 rows x 16 columns]

In [18]:
from sklearn.model_selection import train_test_split

x=df.iloc[:,2:14].values
y=df.iloc[:,-2].values

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)
print(f'X_train shape: {X_train.shape}')
print(f'X_test shape: {X_test.shape}')
print(f'y_train shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')

X_train shape: (20671, 12)
X_test shape: (8860, 12)
y_train shape: (20671,)
y_test shape: (8860,)


In [21]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


model_linear = LinearRegression()
model_linear.fit(X_train, y_train)

y_pred_linear = model_linear.predict(X_test)

mse_linear = mean_squared_error(y_test, y_pred_linear)
mae_linear = mean_absolute_error(y_test, y_pred_linear)
r2_linear = r2_score(y_test, y_pred_linear)
rmse_linear = mse_linear ** 0.5

print(f"Linear Regression MSE: {mse_linear}")
print(f"Linear Regression MAE: {mae_linear}")
print(f"Linear Regression RMSE: {rmse_linear}")
print(f"Linear Regression R² Score: {r2_linear}")

Linear Regression MSE: 1.735979448002134e-05
Linear Regression MAE: 0.0031087309638913223
Linear Regression RMSE: 0.004166508667940262
Linear Regression R² Score: 0.9999999983400809


Linear Regression Metrics:

Mean Squared Error (MSE):
Value: 1.735979448002134
Interpretation: This is a very low MSE, indicating that the predictions are very close to the actual values on average. The model's error is quite small.

Mean Absolute Error (MAE):
Value: 0.0031087309638913223
Interpretation: This low MAE suggests that the average error of the predictions is approximately 0.0031 units. The model is making very accurate predictions.

Root Mean Squared Error (RMSE):
Value: 0.004166508667940262
Interpretation: The RMSE is also very low, indicating that the average distance between the predicted and actual values is minimal.

R² Score:
Value: 0.9999999983400809
Interpretation: This value is very close to 1, suggesting that the linear regression model explains nearly all the variability in the data. It indicates an excellent fit.

In [24]:
from sklearn.neighbors import KNeighborsRegressor

model_knn = KNeighborsRegressor(n_neighbors=7)
model_knn.fit(X_train, y_train)

y_pred_knn = model_knn.predict(X_test)

mse_knn = mean_squared_error(y_test, y_pred_knn)
mae_knn = mean_absolute_error(y_test, y_pred_knn)
r2_knn = r2_score(y_test, y_pred_knn)
rmse_knn = mse_knn ** 0.5

print(f"KNN Regression MSE: {mse_knn}")
print(f"KNN Regression MAE: {mae_knn}")
print(f"KNN Regression RMSE: {rmse_knn}")
print(f"KNN Regression R² Score: {r2_knn}")


KNN Regression MSE: 203.17986030520112
KNN Regression MAE: 7.337239438890681
KNN Regression RMSE: 14.254117310629976
KNN Regression R² Score: 0.9805722281420379


Mean Squared Error (MSE):
Value: 203.17986030520112
Interpretation: This MSE is significantly higher than that of the Linear Regression model, indicating that the KNN model's predictions are, on average, much further from the actual values.

Mean Absolute Error (MAE):
Value: 7.337239438890681
Interpretation: This value shows a higher average prediction error compared to Linear Regression, suggesting that the KNN model may struggle with some predictions.

Root Mean Squared Error (RMSE):
Value: 14.254117310629976
Interpretation: This RMSE is quite large, indicating that the KNN model’s predictions are on average considerably far from the actual values.

R² Score:
Value: 0.9805722281420379
Interpretation: While still a good score, this R² value suggests that the KNN model explains only about 98.06% of the variance in the data, which is less than the linear regression model.

Overall Comparison:
Performance: The Linear Regression model significantly outperforms the KNN model in this case. The very low MSE, MAE, and RMSE values for Linear Regression suggest that it captures the underlying data relationships much more effectively than KNN.

Interpretation of R²: The R² value indicates that Linear Regression fits the data almost perfectly, while KNN shows good but not as high performance.

Conclusion:
In this particular scenario, Linear Regression is the superior model for predicting AQI based on the metrics calculated. KNN may not be suitable for this data, especially if the underlying relationships are more linear. You may want to explore hyperparameter tuning for KNN or even consider other regression techniques if you're looking for improvements in its performance.